File: figma_helper.py
#!/usr/bin/env python3
"""
Figma API Helper
Использование: python figma_helper.py [file_id] [node_id]
"""

import sys
import json
import urllib.request
import urllib.parse

FIGMA_TOKEN = 'figd_KuMWwJhQFHbvwuqjVMaG_F0Eo1shno_fAPpOWf-N'

def make_figma_request(url):
    """Выполняет запрос к Figma API"""
    try:
        req = urllib.request.Request(url)
        req.add_header('X-Figma-Token', FIGMA_TOKEN)
        
        with urllib.request.urlopen(req) as response:
            data = response.read().decode('utf-8')
            return json.loads(data)
    except Exception as e:
        print(f"❌ Ошибка запроса: {e}")
        return None

def get_figma_file(file_id, node_id=None):
    """Получает данные файла или узла из Figma"""
    if node_id:
        url = f"https://api.figma.com/v1/files/{file_id}/nodes?ids={node_id}"
    else:
        url = f"https://api.figma.com/v1/files/{file_id}"
    
    return make_figma_request(url)

def export_figma_image(file_id, node_id, format='png', scale=2):
    """Экспортирует изображение из Figma"""
    url = f"https://api.figma.com/v1/images/{file_id}?ids={node_id}&format={format}&scale={scale}"
    return make_figma_request(url)

def analyze_figma_structure(data, level=0):
    """Анализирует структуру Figma файла"""
    indent = "  " * level
    
    if isinstance(data, dict):
        if 'name' in data and 'type' in data:
            node_info = f"{indent}📄 {data['type']}: {data['name']}"
            if 'id' in data:
                node_info += f" (ID: {data['id']})"
            print(node_info)
            
            if 'children' in data:
                for child in data['children']:
                    analyze_figma_structure(child, level + 1)

def main():
    """Основная функция"""
    args = sys.argv[1:]
    file_id = args[0] if args else 'xMgRBJN5wvmvB7s0miwHmM'
    node_id = args[1] if len(args) > 1 else None
    
    print(f"🔍 Анализируем Figma файл: {file_id}")
    
    if node_id:
        print(f"🎯 Узел: {node_id}")
        
        # Получаем данные узла
        node_data = get_figma_file(file_id, node_id)
        if node_data:
            print("📄 Данные узла:")
            print(json.dumps(node_data, indent=2, ensure_ascii=False))
            
            # Экспортируем изображение
            print("\n🖼️  Экспортируем изображение...")
            image_data = export_figma_image(file_id, node_id)
            if image_data and 'images' in image_data:
                image_url = image_data['images'].get(node_id)
                if image_url:
                    print(f"🔗 Ссылка на изображение: {image_url}")
                else:
                    print("❌ Не удалось получить ссылку на изображение")
    else:
        # Получаем весь файл
        file_data = get_figma_file(file_id)
        if file_data:
            print(f"📁 Файл: {file_data.get('name', 'Без названия')}")
            print(f"📅 Версия: {file_data.get('version', 'Неизвестно')}")
            print(f"📊 Последнее изменение: {file_data.get('lastModified', 'Неизвестно')}")
            
            if 'document' in file_data:
                print("\n📋 Структура:")
                analyze_figma_structure(file_data['document'])

if __name__ == '__main__':
    main()

File: figma-helper.js
// Figma API Helper
// Использование: node figma-helper.js [fileId] [nodeId]

const https = require('https');
const fs = require('fs');

const FIGMA_TOKEN = 'figd_KuMWwJhQFHbvwuqjVMaG_F0Eo1shno_fAPpOWf-N';

async function getFigmaFile(fileId, nodeId = null) {
    return new Promise((resolve, reject) => {
        const url = nodeId 
            ? `https://api.figma.com/v1/files/${fileId}/nodes?ids=${nodeId}`
            : `https://api.figma.com/v1/files/${fileId}`;
            
        const options = {
            headers: {
                'X-Figma-Token': FIGMA_TOKEN
            }
        };

        https.get(url, options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const result = JSON.parse(data);
                    resolve(result);
                } catch (e) {
                    reject(e);
                }
            });
        }).on('error', reject);
    });
}

async function exportFigmaImage(fileId, nodeId, format = 'png', scale = 2) {
    return new Promise((resolve, reject) => {
        const url = `https://api.figma.com/v1/images/${fileId}?ids=${nodeId}&format=${format}&scale=${scale}`;
        
        const options = {
            headers: {
                'X-Figma-Token': FIGMA_TOKEN
            }
        };

        https.get(url, options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    const result = JSON.parse(data);
                    resolve(result);
                } catch (e) {
                    reject(e);
                }
            });
        }).on('error', reject);
    });
}

// Основная функция
async function main() {
    const args = process.argv.slice(2);
    const fileId = args[0] || 'xMgRBJN5wvmvB7s0miwHmM';
    const nodeId = args[1];

    try {
        console.log('🔍 Получаем данные из Figma...');
        
        if (nodeId) {
            // Получаем конкретный узел
            const nodeData = await getFigmaFile(fileId, nodeId);
            console.log('📄 Данные узла:', JSON.stringify(nodeData, null, 2));
            
            // Экспортируем изображение
            console.log('🖼️  Экспортируем изображение...');
            const imageData = await exportFigmaImage(fileId, nodeId);
            console.log('🔗 Ссылка на изображение:', imageData.images[nodeId]);
        } else {
            // Получаем весь файл
            const fileData = await getFigmaFile(fileId);
            console.log('📁 Структура файла:');
            console.log('- Название:', fileData.name);
            console.log('- Версия:', fileData.version);
            console.log('- Страницы:', fileData.document.children.map(page => ({
                name: page.name,
                id: page.id,
                children: page.children?.length || 0
            })));
        }
    } catch (error) {
        console.error('❌ Ошибка:', error.message);
    }
}

if (require.main === module) {
    main();
}

module.exports = { getFigmaFile, exportFigmaImage };














































































